okpt/diff/diff.py

```diff
@@ -0,0 +1,141 @@
+    """
+
+    def __init__(self, base_result: Dict[str, Any],
+                 changed_result: Dict[str, Any], metadata: bool):
+        """Initializes test results and validate them."""
+        self.base_result = base_result
+        self.changed_result = changed_result
+        self.metadata = metadata
+
+        # validate test result keys
+        is_valid, key, result = self._validate_keys()
+        if not is_valid:
+            raise InvalidTestResult(key, result)
+
+        self.base_results = self.base_result['results']
+        self.changed_results = self.changed_result['results']
+
+        # validate test result structure and types
+        is_valid, key, result = self._validate_structure()
+        if not is_valid:
+            raise InconsistentTestResultsError(key, result)
+        is_valid, key, result = self._validate_types()
+        if not is_valid:
+            raise InvalidTestResultType(key, result)
+
+    def _validate_keys(self) -> Tuple[bool, str, str]:
+        """Ensure both test results have `metadata` and `results` keys."""
+        check_keydict = lambda key, res: key in res and isinstance(
+            res[key], dict)
+        if self.metadata:
+            if not check_keydict('metadata', self.base_result):
+                return (False, 'metadata', 'base_result')
+            if not check_keydict('metadata', self.changed_result):
+                return (False, 'metadata', 'changed_result')
+        if not check_keydict('results', self.base_result):
+            return (False, 'results', 'base_result')
+        if not check_keydict('results', self.changed_result):
+            return (False, 'results', 'changed_result')
+        return (True, '', '')
+
+    def _validate_structure(self) -> Tuple[bool, str, str]:
+        """Ensure both test results have the same keys."""
+        for k in self.base_results:
+            if not k in self.changed_results:
+                return (False, k, 'changed_result')
+        for k in self.changed_results:
+            if not k in self.base_results:
+                return (False, k, 'base_result')
+        return (True, '', '')
+
+    def _validate_types(self) -> Tuple[bool, str, str]:
+        """Ensure both test results have numeric values."""
+        for k, v in self.base_results.items():
+            if not _is_numeric(v):
+                return (False, k, 'base_result')
+        for k, v in self.changed_results.items():
+            if not _is_numeric(v):
+                return (False, k, 'changed_result')
+        return (True, '', '')
+
+    def diff(self) -> Dict[str, Any]:
+        """Return the diff between the two test results. (changed - base)"""
+        results_diff = {
+            key: self.changed_results[key] - self.base_results[key]
+            for key in self.base_results
+        }
+
+        # add metadata if specified
+        if self.metadata:
+            return {
+                'base_metadata': self.base_result['metadata'],
```

Comment:
For future PR: What is python's recommendation on handling constants? I see we have a lot of constants that are hardcoded, making it difficult to change them in the future.

Reply:
The Google styleguide [encourages](https://google.github.io/styleguide/pyguide.html#3164-guidelines-derived-from-guidos-recommendations) having module level constants in upper snake case `CONSTANT_NAME`. In this case, they would be private (`_PRIVATE_CONSTANT_NAME`). pep8 [encourages](https://www.python.org/dev/peps/pep-0008/#constants) this as well.

Reply:
I meant more along the lines of having hard coded strings duplicated in multiple places of the code. This makes it trickier to change a single string. I am wondering if it is better to consolidate these constants to variables that can be reused.

Reply:
I've refactored these constants with `Enums` in the latest commit. Is this sort of what you had in mind?

Reply:
I think thats good.